docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,147 @@
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan-optional)
+    - [Graduation Criteria](#graduation-criteria-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+- **Azure Service Operator (ASO)** - ([link](https://azure.github.io/azure-service-operator/)) a collection of Kubernetes CRDs and controllers which extend the Kubernetes API to manage Azure resources like virtual machines and resource groups.
+
+## Summary
+
+This proposal describes how CAPZ could utilize Azure Service Operator (ASO) to create, read, update, and delete Azure resources with a Kubernetes-native interface.
+
+## Motivation
+
+CAPZ currently interfaces with the Azure APIs using https://github.com/Azure/azure-sdk-for-go, a thin wrapper over Azure's REST API, to manage Azure resources backing Cluster API clusters. The Azure Go SDK has recently deprecated its set of APIs that CAPZ uses in favor of a new set of backwards-incompatible APIs. Because the deprecated Azure Go SDK APIs are no longer receiving updates to adopt new Azure API versions, CAPZ must adopt the new APIs to be able to expose the newest features in Azure to Cluster API users.
+
+Because ASO already uses the new APIs, adopting ASO in CAPZ accomplishes the goal of migrating off of the deprecated APIs. In addition, the declarative, Kubernetes-native interface exposed by ASO unlocks opportunities to simplify how CAPZ reconciles resources in Azure.
+
+### Goals
+
+- Adopt ASO in CAPZ to handle creating, updating, and deleting Azure resources
+- Determine a method to migrate CAPZ's individual Azure service interfaces one-by-one to use ASO
+- Determine how ASO will be installed on Cluster API management clusters
+
+### Non-Goals/Future Work
+
+- Entirely eliminate direct usage of any Azure Go SDK
+- Handle changes to Azure resources asynchronously
+- Allow users to interact with ASO resources directly, out-of-band with CAPZ
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+TODO
+
+### Requirements (Optional)
+
+#### Functional Requirements
+
+##### FR1
+
+TODO
+
+#### Non-Functional Requirements
+
+##### NFR1
+
+TODO
+
+### Implementation Details/Notes/Constraints
+
+- ASO needs to be installed on the management cluster
+  - Multi- or single-tenant?
+  - Install in capz-controller-manager `func main()`
+- GET requests are not as straightforward in the ASO model because BYO resources
+  may not initially exist in ASO
+- Current service pattern in CAPZ works for Azure services that are not
+  represented in ASO (resource health)
+- Each release of ASO can expose several Azure API versions of each resource,
+  including GA and preview API versions.
+
+### Security Model
+
+- ASO resources are subject to Kubernetes RBAC => capz-controller-manager will require many more permissions that may extend beyond what's minimally required.
```

Comment:
we should also talk how to handle creds here (or at least put it as a TODO for now)